analytics/project_tracker.go

```diff
@@ -0,0 +1,44 @@
+package analytics
+
+import "sync"
+
+type ProjectTracker struct {
+	earthfileOrg     string
+	earthfileProject string
+	cliOrg           string
+	cliProject       string
+	mutex            sync.Locker
+}
+
+var projectTracker = ProjectTracker{
+	mutex: &sync.Mutex{},
+}
+
+func (pt *ProjectTracker) AddEarthfileProject(org, project string) {
+	pt.mutex.Lock()
+	defer pt.mutex.Unlock()
+	pt.earthfileOrg = org
+	pt.earthfileProject = project
+}
+
+func (pt *ProjectTracker) AddCLIProject(org, project string) {
+	pt.cliOrg = org
+	pt.cliProject = project
+}
+
+func (pt *ProjectTracker) ProjectDetails() (string, string) {
+	if pt.earthfileOrg != "" && pt.earthfileProject != "" {
+		return pt.earthfileOrg, pt.earthfileProject
+	}
+	org := pt.cliOrg
+	project := pt.cliProject
+	pt.mutex.Lock()
```

Comment:
What I had in mine here is that I can avoid lock/unlock if the two values are already populated,
and then if either one of them is not, the lock is acquired and we check again if the values are populated/empty.

I don't have a problem to move it to the top and avoid checking the values twice if you find it unnecessary.

Reply:
> What I had in mine here is that I can avoid lock/unlock if the two values are already populated,

But you need a memory barrier to occur, otherwise you might have one thread whos CPU has updated the ram memory; however another thread on a completely different cpu (with a completely different L1/L2 cache) might return a cached copy of the memory without being forced to check the value stored in ram.

one could omit the lock in this case if there were a call to https://man7.org/linux/man-pages/man2/membarrier.2.html ; however, moving the Lock() call higher up would achieve this while being following standard go idioms.

Reply:
discussed offline, moved lock to the top of the function